src/ast/irbuilderbpf.cpp

```diff
@@ -819,6 +819,146 @@ Value *IRBuilderBPF::CreateStrncmp(Value *str1,
 
   return result;
 }
+Value *IRBuilderBPF::CreateStrstr(Value *val1,
+                                  uint64_t str1_size,
+                                  Value *val2,
+                                  uint64_t str2_size,
+                                  bool inverse)
+{
+  /*
+  // This function compares whether the string val1 contains the string val2.
+  // It returns true if val2 is contained by val1, false if not contained.
+  // strstr(String val1, String val2, int str1_size, int str2_size)
+     {
+        for (size_t j = 0; (str1_size >= str2_size) && (j <= str1_size -
+  str2_size); j++)
+        {
+          for (size_t i = 0; i < str2_size; i++)
+          {
+            if (val2[i] == NULL)
+            {
+              return true;
+            }
+            if (val1[i + j] != val2[i])
+            {
+              break;
+            }
+          }
+          if (val1[j] == NULL) {
+            return false;
+          }
+        }
+        return false;
+     }
+  */
+  // Check if the compared strings are literals.
+  // If so, we can avoid storing the literal in memory.
+  std::optional<std::string> literal1;
+  if (auto constString1 = dyn_cast<ConstantDataArray>(val1))
+    literal1 = constString1->getAsString();
+  else if (isa<ConstantAggregateZero>(val1))
+    literal1 = "";
+  else
+    literal1 = std::nullopt;
+
+  std::optional<std::string> literal2;
+  if (auto constString2 = dyn_cast<ConstantDataArray>(val2))
+    literal2 = constString2->getAsString();
+  else if (isa<ConstantAggregateZero>(val2))
+    literal2 = "";
+  else
+    literal2 = std::nullopt;
+
+  Function *parent = GetInsertBlock()->getParent();
+  AllocaInst *store = CreateAllocaBPF(getInt1Ty(), "strstr.result");
+  BasicBlock *done_true = BasicBlock::Create(module_.getContext(),
+                                             "strstr.true",
+                                             parent);
+  BasicBlock *done_false = BasicBlock::Create(module_.getContext(),
+                                              "strstr.false",
+                                              parent);
+
+  CreateStore(getInt1(inverse), store);
+
+  Value *null_byte = getInt8(0);
+  for (size_t j = 0; (str1_size >= str2_size) && (j <= str1_size - str2_size);
+       j++)
+  {
+    BasicBlock *one_loop = BasicBlock::Create(module_.getContext(),
+                                              "one.loop",
+                                              parent);
+
+    Value *str_c;
+    if (literal1)
+      str_c = getInt8(literal1->c_str()[j]);
+    else
+    {
+      auto *ptr_str = CreateGEP(ArrayType::get(getInt8Ty(), str1_size),
+                                val1,
+                                { getInt32(0), getInt32(j) });
+      str_c = CreateLoad(getInt8Ty(), ptr_str);
+    }
+
+    for (size_t i = 0; i < str2_size; i++)
+    {
+      BasicBlock *two_loop = BasicBlock::Create(module_.getContext(),
+                                                "two.loop",
```

Comment:
Basic block names would ideally be unique. Can you append the idx in the first string here?